// learnpack/src/managers/config/index.js
const path = require("path");
const fs = require("fs");
const shell = require("shelljs");
const Console = require("../../utils/console");
const watch = require("../../utils/watcher");
const chalk = require("chalk");
const fetch = require("node-fetch");
const {
  ValidationError,
  NotFoundError,
  InternalError,
} = require("../../utils/errors.js");

let defaults = require("./defaults.js");
let { exercise } = require("./exercise.js");

const { rmSync } = require("../file.js");
/* exercise folder name standard */

const getConfigPath = () => {
  const possibleFileNames = [
    "learn.json",
    ".learn/learn.json",
    "bc.json",
    ".breathecode/bc.json",
  ];
  let config = possibleFileNames.find((file) => fs.existsSync(file)) || null;
  if (config && fs.existsSync(".breathecode"))
    return { config, base: ".breathecode" };
  else if (config === null)
    throw NotFoundError(
      "learn.json file not found on current folder, is this a learnpack package?"
    );
  return { config, base: ".learn" };
};

const getExercisesPath = (base) => {
  const possibleFileNames = ["./exercises", base + "/exercises", "./"];
  return possibleFileNames.find((file) => fs.existsSync(file)) || null;
};

const getGitpodAddress = () => {
  if (shell.exec(`gp -h`, { silent: true }).code == 0) {
    return shell
      .exec(`gp url`, { silent: true })
      .stdout.replace(/(\r\n|\n|\r)/gm, "");
  } else {
    Console.debug(`Gitpod command line tool not found`);
    return "http://localhost";
  }
};

module.exports = async ({ grading, mode, disableGrading, version }) => {
  let confPath = getConfigPath();
  Console.debug("This is the config path: ", confPath);

  let configObj = {};
  if (confPath) {
    const bcContent = fs.readFileSync(confPath.config);

    let hiddenBcContent = {};
    if (fs.existsSync(confPath.base + "/config.json")) {
      hiddenBcContent = fs.readFileSync(confPath.base + "/config.json");
      hiddenBcContent = JSON.parse(hiddenBcContent);
      if (!hiddenBcContent)
        throw Error(
          `Invalid ${confPath.base}/config.json syntax: Unable to parse.`
        );
    }

    const jsonConfig = JSON.parse(bcContent);
    if (!jsonConfig)
      throw Error(`Invalid ${confPath.config} syntax: Unable to parse.`);

    //add using id to the installation
    if (!jsonConfig.session)
      jsonConfig.session = Math.floor(Math.random() * 10000000000000000000);

    configObj = deepMerge(hiddenBcContent, { config: jsonConfig }, {
      config: { disableGrading },
    });
    Console.debug("Content form the configuration .json ", configObj);
  } else {
    throw ValidationError(
      "No learn.json file has been found, make sure you are in the folder"
    );
  }

  configObj = deepMerge(defaults || {}, configObj, {
    config: {
      grading: grading || configObj.grading,
      configPath: confPath.config,
    },
  });
  configObj.config.outputPath = confPath.base + "/dist";

  Console.debug("This is your configuration object: ", {
    ...configObj,
    exercises: configObj.exercises
      ? configObj.exercises.map((e) => e.slug)
      : [],
  });

  // auto detect agent (if possible)
  if (shell.which("gp")) {
    configObj.config.editor.agent = "gitpod";
    configObj.config.address = getGitpodAddress();
    configObj.config.publicUrl = `https://${
      configObj.config.port
    }-${configObj.config.address.substring(8)}`;
  } else if (!configObj.config.editor.agent) {
    configObj.config.editor.agent = "localhost";
  }

  if (!configObj.config.publicUrl)
    configObj.config.publicUrl = `${configObj.config.address}:${configObj.config.port}`;

  // Assign default editor mode if not set already
  if (mode != null) {
    configObj.config.editor.mode = mode;
  }

  if (!configObj.config.mode)
    configObj.config.editor.mode =
      configObj.config.editor.agent === "localhost" ? "standalone" : "preview";

  if (version) configObj.config.editor.version = version;
  else if (configObj.config.editor.version === null) {
    const resp = await fetch(
      "https://raw.githubusercontent.com/learnpack/coding-ide/learnpack/package.json"
    );
    const packageJSON = await resp.json();
    configObj.config.editor.version = packageJSON.version || "1.0.61";
  }

  configObj.config.dirPath = "./" + confPath.base;
  configObj.config.exercisesPath = getExercisesPath(confPath.base) || "./";

  return {
    validLanguages: {},
    get: () => configObj,
    validateEngine: function (language, server, socket) {
      const alias = (_l) => {
        let map = {
          python3: "python",
        };
        if (map[_l]) return map[_l];
        else return _l;
      };

      // decode aliases
      language = alias(language);

      if (this.validLanguages[language]) return true;

      Console.debug(`Validating engine for ${language} compilation`);
      let result = shell.exec("learnpack plugins", { silent: true });

      if (result.code == 0 && result.stdout.includes(`learnpack-${language}`)) {
        this.validLanguages[language] = true;
        return true;
      }

      Console.info(`Language engine for ${language} not found, installing...`);
      result = shell.exec(`learnpack plugins:install learnpack-${language}`, {
        silent: true,
      });
      if (result.code === 0) {
        socket.log(
          "compiling",
          "Installing the python compiler, you will have to reset the exercises after installation by writing on your terminal: $ learnpack run"
        );
        Console.info(
          `Successfully installed the ${language} exercise engine, \n please start learnpack again by running the following command: \n ${chalk.white(
            `$ learnpack start`
          )}\n\n `
        );
        server.terminate();
        return false;
      } else {
        this.validLanguages[language] = false;
        socket.error(`Error installing ${language} exercise engine`);
        Console.error(`Error installing ${language} exercise engine`);
        Console.log(result.stdout);
        throw InternalError(`Error installing ${language} exercise engine`);
      }
    },
    clean: () => {
      rmSync(configObj.config.outputPath);
      rmSync(configObj.config.dirPath + "/_app");
      rmSync(configObj.config.dirPath + "/reports");
      rmSync(configObj.config.dirPath + "/.session");
      rmSync(configObj.config.dirPath + "/resets");

      // clean tag gz
      if (fs.existsSync(configObj.config.dirPath + "/app.tar.gz"))
        fs.unlinkSync(configObj.config.dirPath + "/app.tar.gz");

      if (fs.existsSync(configObj.config.dirPath + "/config.json"))
        fs.unlinkSync(configObj.config.dirPath + "/config.json");

      if (fs.existsSync(configObj.config.dirPath + "/vscode_queue.json"))
        fs.unlinkSync(configObj.config.dirPath + "/vscode_queue.json");
    },
    getExercise: (slug) => {
      const exercise = configObj.exercises.find((ex) => ex.slug == slug);
      if (!exercise) throw ValidationError(`Exercise ${slug} not found`);

      return exercise;
    },
    getAllExercises: () => {
      return configObj.exercises;
    },
    startExercise: function (slug) {
      const exercise = this.getExercise(slug);

      // set config.json with current exercise
      configObj.currentExercise = exercise.slug;

      this.save();

      exercise.files.forEach((f) => {
        const _path = configObj.config.outputPath + "/" + f.name;
        if (f.hidden === false && fs.existsSync(_path)) fs.unlinkSync(_path);
      });

      return exercise;
    },
    noCurrentExercise: function () {
      configObj.currentExercise = null;
      this.save();
    },
    reset: (slug) => {
      if (!fs.existsSync(`${configObj.config.dirPath}/resets/` + slug))
        throw ValidationError("Could not find the original files for " + slug);

      const exercise = configObj.exercises.find((ex) => ex.slug == slug);
      if (!exercise)
        throw ValidationError(
          `Exercise ${slug} not found on the configuration`
        );

      fs.readdirSync(`${configObj.config.dirPath}/resets/${slug}/`).forEach(
        (fileName) => {
          const content = fs.readFileSync(
            `${configObj.config.dirPath}/resets/${slug}/${fileName}`
          );
          fs.writeFileSync(`${exercise.path}/${fileName}`, content);
        }
      );
    },
    buildIndex: function () {
      Console.info("Building the exercise index...");

      const isDirectory = (source) => {
        const name = path.basename(source);
        if (name === path.basename(configObj.config.dirPath)) return false;
        //ignore folders that start with a dot
        if (name.charAt(0) === "." || name.charAt(0) === "_") return false;

        return fs.lstatSync(source).isDirectory();
      };
      const getDirectories = (source) =>
        fs
          .readdirSync(source)
          .map((name) => path.join(source, name))
          .filter(isDirectory);
      // add the .learn folder
      if (!fs.existsSync(confPath.base)) fs.mkdirSync(confPath.base);
      // add the outout folder where webpack will publish the the html/css/js files
      if (
        configObj.config.outputPath &&
        !fs.existsSync(configObj.config.outputPath)
      )
        fs.mkdirSync(configObj.config.outputPath);

      // TODO: we could use npm library front-mater to read the title of the exercises from the README.md
      const grupedByDirectory = getDirectories(configObj.config.exercisesPath);
      if (grupedByDirectory.length > 0)
        configObj.exercises = grupedByDirectory.map((path, position) =>
          exercise(path, position, configObj)
        );
      // else means the exercises are not in a folder
      else
        configObj.exercises = [
          exercise(configObj.config.exercisesPath, 0, configObj),
        ];
      this.save();
    },
    watchIndex: function (onChange = null) {
      if (!configObj.config.exercisesPath)
        throw ValidationError(
          "No exercises directory to watch: " + configObj.config.exercisesPath
        );

      this.buildIndex();
      watch(configObj.config.exercisesPath)
        .then((eventname, filename) => {
          Console.debug("Changes detected on your exercises");
          this.buildIndex();
          if (onChange) onChange();
        })
        .catch((error) => {
          throw error;
        });
    },
    save: (config = null) => {
      Console.debug("Saving configuration with: ", configObj);

      //remove the duplicates form the actions array
      // configObj.config.actions = [...new Set(configObj.config.actions)];
      configObj.config.translations = [
        ...new Set(configObj.config.translations),
      ];

      fs.writeFileSync(
        configObj.config.dirPath + "/config.json",
        JSON.stringify(configObj, null, 4)
      );
    },
  };
};

function deepMerge(...sources) {
  let acc = {};
  for (const source of sources) {
    if (source instanceof Array) {
      if (!(acc instanceof Array)) {
        acc = [];
      }
      acc = [...source];
    } else if (source instanceof Object) {
      for (let [key, value] of Object.entries(source)) {
        if (value instanceof Object && key in acc) {
          value = deepMerge(acc[key], value);
        }
        if (value !== undefined) acc = { ...acc, [key]: value };
      }
    }
  }
  return acc;
}
